### PR TITLE
DOC: Ensure that we add documentation also as to the dict for types

### LINKED ISSUE #18740: dtype missing docstrings since 1.20.1
### Reproducing code example:

```python
import numpy as np
print(np.dtype.__doc__)
```

### Error message:

`None`

Expected:
```
dtype(obj, align=False, copy=False)

    Create a data type object.

    A numpy array is homogeneous, and contains elements described by a
    dtype object. A dtype object can be constructed from different
    combinations of fundamental numeric types.

    Parameters
    ----------
    obj
        Object to be converted to a data type object.
    align : bool, optional
        Add padding to the fields to match what a C compiler would output
        for a similar C-struct. Can be ``True`` only if `obj` is a dictionary
        or a comma-separated string. If a struct dtype is being created,
        this also sets a sticky alignment flag ``isalignedstruct``.
    copy : bool, optional
        Make a new copy of the data-type object. If ``False``, the result
        may just be a reference to a built-in data-type object.

    See also
    --------
    result_type

    Examples
    --------
    Using array-scalar type:

    >>> np.dtype(np.int16)
    dtype('int16')

    Structured type, one field name 'f1', containing int16:

    >>> np.dtype([('f1', np.int16)])
    dtype([('f1', '<i2')])

    Structured type, one field named 'f1', in itself containing a structured
    type with one field:

    >>> np.dtype([('f1', [('f1', np.int16)])])
    dtype([('f1', [('f1', '<i2')])])

    Structured type, two fields: the first field contains an unsigned int, the

    >>> np.dtype([('f1', np.uint64), ('f2', np.int32)])
    dtype([('f1', '<u8'), ('f2', '<i4')])

    Using array-protocol type strings:

    >>> np.dtype([('a','f8'),('b','S10')])
    dtype([('a', '<f8'), ('b', 'S10')])
    Using comma-separated field formats.  The shape is (2,3):

    >>> np.dtype("i4, (2,3)f8")
    dtype([('f0', '<i4'), ('f1', '<f8', (2, 3))])

    Using tuples.  ``int`` is a fixed type, 3 the field's shape.  ``void``

    >>> np.dtype([('hello',(np.int64,3)),('world',np.void,10)])
    dtype([('hello', '<i8', (3,)), ('world', 'V10')])

    Subdivide ``int16`` into 2 ``int8``'s, called x and y.  0 and 1 are
    the offsets in bytes:

    >>> np.dtype((np.int16, {'x':(np.int8,0), 'y':(np.int8,1)}))
    dtype((numpy.int16, [('x', 'i1'), ('y', 'i1')]))

    Using dictionaries.  Two fields named 'gender' and 'age':

    >>> np.dtype({'names':['gender','age'], 'formats':['S1',np.uint8]})
    dtype([('gender', 'S1'), ('age', 'u1')])

    Offsets in bytes, here 0 and 25:

    >>> np.dtype({'surname':('S25',0),'age':(np.uint8,25)})
    dtype([('surname', 'S25'), ('age', 'u1')])
```

### NumPy/Python version information:

import sys, numpy; print(numpy.__version__, sys.version)
1.20.1 3.9.2 (tags/v3.9.2:1a79785, Feb 19 2021, 13:44:55) [MSC v.1928 64 bit (AMD64)]

### PR DESCRIPTION
**Pull request by seberg**
This ensures that `help(np.dtype)` produces a result.  I am not
exactly sure why it picks up `__doc__` from the dict instead of
`tp_doc` right now. It probably is due to the combination of
inheritance and the fact that the dict always includes `None`
and gets preference during inheritance.
(That probably makes a lot of sense to not inherit the `type`
docstring by default.)

Modifying the dictionary directly is not really good style, either,
but hopefully works.

Closes gh-18740

### CODE REVIEW

**Comment by seberg** [comment]
OK, this doesn't quite seem to work, because it overshadows the getset of the instance? Hmmm...

**Comment by eric-wieser** [comment]
Are metaclasses related at all here, or is this just a heap type problem?

**Comment by seberg** [comment]
@eric-wieser yeah, that would make sense, but I am not setting the HEAPTYPE flag on `np.dtype` right now, which is why I am confused.  So my guess was that inheritance causes the `__dict__` value to be used instead of the `tp_doc` one (although I am not quite sure where that happens in pythons `typeobject.c`

**Comment by charris** [comment]
close/reopen

**Comment by seberg** [comment]
After chasing around how ufuncs add docstrings (we have a function to do it via Python, which claims it is used but is not actually used)...

The only thing that was necessary here was to only put the doc-string into the dict only if the dict contains `None` for `__doc__` (which means `PyType_Read()` probably filled it in automatically.

This actually means/fixes that `np.float16(3.).__doc__` also correctly shows the `np.float16.__doc__` now.

File: numpy/core/src/multiarray/compiled_base.c [hunk 1]
```diff
@@ -1428,9 +1428,26 @@ arr_add_docstring(PyObject *NPY_UNUSED(dummy), PyObject *args)
...
-    else if (Py_TYPE(obj) == &PyType_Type) {
+    else if (PyObject_TypeCheck(obj, &PyType_Type)) {
+        /*
+         * We add it to both `tp_doc` and `__doc__` here.  Note that in theory
+         * `tp_doc` extracts the signature line, but we currently do not use
+         * it.  It may make sense to only add it as `__doc__` and
+         * `__text_signature__` to the dict in the future.
+         * The dictionary path is only necessary for heaptypes (currently not
+         * used) and metaclasses.
+         * If `__doc__` as stored in `to_dict` is None, we assume this was
```
Comment (WarrenWeckesser) [inline numpy/core/src/multiarray/compiled_base.c:1439]:
```suggestion
         * If `__doc__` as stored in `tp_dict` is None, we assume this was
```

**Review by WarrenWeckesser — APPROVED** [review]
LGTM, thanks @seberg.  I noticed one small typo.

File: numpy/core/src/multiarray/compiled_base.c [hunk 2]
```diff
@@ -1428,9 +1428,26 @@ arr_add_docstring(PyObject *NPY_UNUSED(dummy), PyObject *args)
...
+         * filled in by `PyType_Read()` and should also be replaced.
```
Comment (WarrenWeckesser) [inline numpy/core/src/multiarray/compiled_base.c:1440]: Another typo:
```suggestion
         * filled in by `PyType_Ready()` and should also be replaced.
```

**Pull request merged**